File: 22 - Connect with JS/c. Geolocation - ii. Full [WIP]/wasm-v7/App.razor.cs
// European Union Public License version 1.2
// Copyright © 2023 Rick Beerendonk

using System.Runtime.InteropServices.JavaScript;
using System.Runtime.Versioning;

namespace GeolocationDemo;

public partial class App
{
    [JSImport("getCurrentPosition", "App")]
    [return: JSMarshalAs<JSType.Promise<JSType.Object>>()]
    internal static partial Task<JSObject> GetCurrentPositionJS();

    internal static getCurrentPosition()
    {

    }
}

public class GeolocationPosition
{
  public GeolocationCoordinates Coords { get; set; }
  //   timestamp
 }

public class GeolocationCoordinates
{
    Longitude { get; set; }
    Latitude { get; set; }
    Altitude { get; set; }
    Accuracy { get; set; }
    AltitudeAccuracy { get; set; }
    Heading { get; set; }
    Speed { get; set; }
}

File: 23 - Connect with JS/c. Geolocation - ii. Full [WIP]/wasm-v7/App.razor.js
// European Union Public License version 1.2
// Copyright © 2023 Rick Beerendonk

export async function getCurrentPosition() {
  // MDN Docs: https://developer.mozilla.org/en-US/docs/Web/API/Geolocation/getCurrentPosition

  // pos = {
  //   coords: { longitude, latitude, altitude, accuracy, altitudeAccuracy, heading, speed,  },
  //   timestamp
  // }
  const pos = await new Promise((resolve, reject) => {
    function errorCallback(err) {
      // err = { code, message }
      // code: 1 = PERMISSION_DENIED, 2 = POSITION_UNAVAILABLE, 3 = TIMEOUT
      reject(err.message);
    }

    navigator.geolocation.getCurrentPosition(reject, errorCallback, {
      maximumAge: 0 /* default = 0 */,
      timeout: 10000 /* default = infinity */,
      enableHighAccuracy: true /* default = false, true consumes more energy */,
    });
    reject();
  });

  return pos;
}
